### ISSUE TITLE
How to get meson to not add the -pipe option to my cpp compilation steps

### ISSUE THREAD

**Issue by njpau** [mesonbuild/meson#8508]
I have declared my project settings as such.
`project('MyProject', 'cpp', 'c',
    default_options: ['buildtype=plain', 'b_staticpic=false', 'b_asneeded=false', 'b_lundef=false']`

The project requires cross-compilation and neither my c _args or my cpp_args that come under the properties section of my crossfile include the '-pipe' option.

However, the build.ninja that meson generates for my build directories shows that the '-pipe' has been added as a compiler argument in all my c/cpp compilation steps. Not sure if I missed it, but I'm unable to find a way to not enable meson to do this from the documentation. Is there a default option that I can set to not allow meson to insert compilation arguments other than the ones listed in my crossfile?

System parameters:
OS: Debian stretch
Python version: 3.5.3
Meson version: 0.56.0
Ninja version; 1.7.2

**Comment by eli-schwartz** [comment]
`-pipe` is unconditionally added by meson for compilers that support it. Why do you wish to remove it?

>     -pipe
>         Use pipes rather than temporary files for communication between the various stages of compilation.
>         This fails to work on some systems where the assembler is unable to read from a pipe; but the GNU
>         assembler has no trouble.

**Comment by large2021** [comment]
I'd like to see it removed (or made an option) because it causes OOM compilation failures on some moderately large servers we use as Jenkins agents.  Each agent has 64 cores/128 threads, and 256GB of memory, and about twice a week, we have a job failure due to an OOM caused when the automatically inserted "-pipe" causes more memory usage.  

This causes our users (40+) to interrupt our Jenkins savvy folks (~5) to ask for help when their job dies.

Machines are cheap...people are expensive.  Digging through huge logs is irritating.

"What about adding some swap to prevent the OOM?" you might ask.  I would reply that doing so causes unacceptable slowdowns when the system swaps, and adds another point of failure on machines that largely run out of RAM.  

Bottom line is, we've tuned these machines pretty carefully for turnaround on pipelined build jobs, and don't need to spend manpower diagnosing why jobs randomly fail.

One should be able to turn the generation of "-pipe" on and off, as it is not universally needed, and not always appropriate.  Make it the default, sure, but add some way to disable it, other than hacking meson each release, or post-processing the generated ninja files. I'll accept slightly slower jobs for more robustness, if *I* get to make the decision.

**Comment by njpau** [comment]
Happy to provide a PR that adds functionality for a user to omit the -pipe option, if y'all are happy to have it in.

**Comment by eli-schwartz** [comment]
Om a distro level, `-pipe` is set in the stock `$CFLAGS` (at least for Arch Linux). I'd be fine with leaving this up to users.

**Comment by jpakkane** [comment]
> I'd like to see it removed (or made an option) because it causes OOM compilation failures

I'm curious. How do you know that the culprit is `-pipe` rather than something else?

**Comment by anarazel** [comment]
@njpau It's quite curious that -pipe leads to an OOM. The default maximum pipe buffer size is 1MiB. That's not a whole lot compared to the memory overhead of a compile. Any chance that the defaults for `/proc/sys/fs/pipe-max-size` or `/proc/sys/fs/pipe-user-pages-soft` have drastically been increased?

**Comment by obilaniu** [comment]
@anarazel It's probably not the mere fact of the pipeline's existence causing the OOM, be it 1MB or 16KB, but the simultaneous existence of two or more running processes, such as `cpp | cc1 | as`, each communicating with the next via a pipeline and imposing their own large memory footprint (at least while they live).

In my humble opinion, this is not so much a problem in Meson (a build system that does its part of the job admirably) as it is one of **Ninja** spawning too many processes under high memory pressure. This is a situation that can only really be dynamically detected (ideally), or else must be handled by setting maximum build-parallelism conservatively, sometimes excessively so.

Unfortunately [it appears that](https://github.com/ninja-build/ninja/pull/660#issuecomment-468254247) Ninja upstream [is uninterested in](https://github.com/ninja-build/ninja/pull/1354) a solution on their end. So it seems that idea is at an impasse, unless you are willing to consider the unmaintained patch to Ninja offered by a contributor there?

**Comment by jpakkane** [comment]
One thing you might want to try is to limit the number of linker processes instead (there is a builtin option for it). Linking takes a lot more memory. I have seen OOMs due to the linker fairly often but I don't eve recall seeing one for the compiler.

**Comment by njpau** [comment]
Just to clarify, which builtin option are you referring to?

**Comment by jpakkane** [comment]
`backend_max_links`
